ptypy/custom/threepie.py

```diff
@@ -227,4 +233,30 @@ def multislice_update(self, view):
         for i in range(1, self.p.number_of_slices):
             self.ob *= self._object[i]
 
-        return error
\ No newline at end of file
+        if self.p.object_regularization_rate > 0:
+            self.apply_object_regularization()
+
+        return error
+
+    def apply_object_regularization(self):
+        # single mode implementation
+        # only valide for slices with identical thickness
+        assert(self.p.number_of_slices > 1)
+        assert(isinstance(self.p.slice_thickness, float))
+
+        shape = self._object[0].S["Sscan_00G00"].data.shape[1:]
+        psize = self._object[0].S["Sscan_00G00"].psize[0]
+        kz = np.fft.fftfreq(self.p.number_of_slices, self.p.slice_thickness)[..., np.newaxis, np.newaxis]
+        ky = np.fft.fftfreq(shape[0], psize)[..., np.newaxis]
+        kx = np.fft.fftfreq(shape[1], psize)
+
+        # calculate the weight array
+        w = 1 - 2*np.arctan2(self.p.object_regularization_rate**2 * kz**2, kx**2+ky**2+np.spacing(1))/np.pi
```

Comment:
this part is basically just calculating some weights `w` and does not depend on any current update so can be moved into a separate function, e.g. `initialize_regularizer` and called once in the constructor. In this current implementation the weights are re-calculated for every iteration which seems unnecessary.